README.md
# fastdot
> A simple wrapper over `pydot` to make it more consistent, unsurprising, and pythonic


Acknowledgement: `fastdot` is heavily influenced by work from [David Page](https://github.com/davidcpage/), who built a system for drawing graphs based on a highly flexible data structure he designed.

## Install

We suggest installing with conda: `conda install -c fastai fastdot`. You can alternatively install with pip: `pip install fastdot`; however, if you use this approach, you'll also need to install graphviz (e.g. using `apt`, `brew`, etc).

## Synopsis

Start with some data representing objects and connections between them (e.g. they wouldn't normally be just strings like in this example, but would be neural net layers, or users and products, or car trips, etc):

```python
layers1 = ['conv','conv','lin']
layers2 = ['conv','lin']
block1,block2 = ['block1','block2']
conns = ((block1, block2),
         (block1, layers2[-1]))
```

Then map them directly to a visual respresentation:

```python
g = graph_items(seq_cluster(layers1, block1),
                seq_cluster(layers2, block2))
g.add_items(*object_connections(conns))
g
```




![svg](docs/images/output_7_0.svg)



See the *symbolic graphs* and *object graphs* sections below for a more complete example.

## fastdot overview

`fastdot` is a thin wrapper over the excellent [pydot](https://github.com/pydot/pydot) program (which is in turn a thin wrapper over the absolutely wonderful [Graphviz software](https://www.graphviz.org/)), designed to make it more consistent, unsurprising, and pythonic. (An example of removing *surprise*: `pydot.Node('node')` gives an obscure compilation exception, since `node` is a keyword in the underlying `graphviz` program, whereas `fastdot.Node('node')` works just fine, due to auto-quoting.) In fact, you never need to provide names in `fastdot`; you can create edges directly between objects.

Here's a quick example of some of the main functionality:

```python
g = Dot()
c = Cluster('cl', fillcolor='pink')
a1,a2,b = c.add_items('a', 'a', 'b')
c.add_items(a1.connect(a2), a2.connect(b))
g.add_item(Node('Check tooltip', tooltip="I have a tooltip!"))
g.add_item(c)
g
```




![svg](docs/images/output_12_0.svg)



As you see, graphs know how to show themselves in Jupyter notebooks directly and can be exported to HTML (it uses SVG behind the scenes). Tooltips appear in both notebooks and exported HTML pages. Nodes with the same label, by default, are set to the same color. Also, as shown above, you can just use `add_item` or `add_items`, regardless of the type of item.

## Symbolic graphs

`fastdot` is particularly designed to make it easier to create graphs symbolically - for instance, for Python dictionaries, PyTorch/TensorFlow models, and so forth. Here's a simple example with some mock neural network layers and sequential models. First, let's define our mock classes:

```python
@dataclass(frozen=True)
class Layer: name:str; n_filters:int=1
class Linear(Layer): pass
class Conv2d(Layer): pass

@dataclass(frozen=True)
class Sequential: layers:list; name:str
```

Here's our sequential blocks for our "model":

```python
block1 = Sequential([Conv2d('conv', 5), Linear('lin', 3)], 'block1')
block2 = Sequential([Conv2d('conv1', 8), Conv2d('conv2', 2), Linear('lin')], 'block2')
```

`fastdot` can create all node properties directly from objects; you just have to define functions describing how to map the object's attributes to graph properties. These mappings go in the `node_defaults` and `cluster_defaults` dictionaries (although by default labels are set using `str()`, so we don't need any special cluster defaults in this case):

```python
node_defaults['fillcolor'] = lambda o: 'greenyellow' if isinstance(o,Linear) else 'pink'
cluster_defaults['label'] = node_defaults['label'] = attrgetter('name')
node_defaults['tooltip'] = str
```

With that in place, we can directly create nodes from our objects, for instance using the convenient `seq_cluster` function:

```python
c1 = seq_cluster(block1.layers, block1)
c2 = seq_cluster(block2.layers, block2)
e1,e2 = c1.connect(c2),c1.connect(c2.last())
graph_items(c1,c2,e1,e2)
```




![svg](docs/images/output_22_0.svg)



Note that in this example we didn't even need to create the `Dot` object separately - `graph_items` creates it directly from the graph items provided.

## Using object graphs

In the above example, we defined our edges directly between `fastdot` objects. In practice, however, you'll most likely have your edges defined directly between python objects, for instance like this:

```python
conns = (
    (block1, block2),
    (block1, block2.layers[-1]),
)
```

In this case, you'll want some way to connect your python objects to the `fastdot` graph items that represent them. A mapping is stored automatically by `fastdot`, and is made available through the `object2graph` function:

```python
g = graph_items(seq_cluster(block1.layers, block1), seq_cluster(block2.layers, block2))
object2graph(block1.layers[-1])
```




    <pydot.Node at 0x7fa459e92690>



You can use this to graph your connections without needing access to the graph items:

```python
g.add_items(*[object2graph(a).connect(object2graph(b))
              for a,b in conns])
g
```




![svg](docs/images/output_30_0.svg)



There's a helper function, `object_connections`, which creates these connections for you. So the above can be simplified to:

```python
g = graph_items(seq_cluster(block1.layers, block1), seq_cluster(block2.layers, block2))
g.add_items(*object_connections(conns))
g
```




![svg](docs/images/output_32_0.svg)




CHANGELOG.md
# Release notes

<!-- do not remove -->

## 0.1.2

### Bugs squashed

- Fix missing imports



fastdot/_nbdev.py
# AUTOGENERATED BY NBDEV! DO NOT EDIT!

__all__ = ["index", "modules", "custom_doc_links", "git_url"]

index = {"Dot": "00_core.ipynb",
         "uniq_name": "00_core.ipynb",
         "quote": "00_core.ipynb",
         "graph_objects": "00_core.ipynb",
         "object_names": "00_core.ipynb",
         "add_mapping": "00_core.ipynb",
         "node_defaults": "00_core.ipynb",
         "Node": "00_core.ipynb",
         "object2graph": "00_core.ipynb",
         "obj2node_color": "00_core.ipynb",
         "graph_colors1": "00_core.ipynb",
         "graph_colors2": "00_core.ipynb",
         "cluster_defaults": "00_core.ipynb",
         "Cluster": "00_core.ipynb",
         "pydot.Graph.nodes": "00_core.ipynb",
         "pydot.Graph.__getitem__": "00_core.ipynb",
         "pydot.Graph.add_item": "00_core.ipynb",
         "pydot.Graph.add_items": "00_core.ipynb",
         "graph_items": "00_core.ipynb",
         "pydot.Graph.first": "00_core.ipynb",
         "pydot.Graph.last": "00_core.ipynb",
         "pydot.Node.with_compass": "00_core.ipynb",
         "pydot.Graph.with_compass": "00_core.ipynb",
         "pydot.Node.connect": "00_core.ipynb",
         "pydot.Graph.connect": "00_core.ipynb",
         "object_connections": "00_core.ipynb",
         "graph_edges_seq": "00_core.ipynb",
         "pydot.Graph.add_edges_seq": "00_core.ipynb",
         "seq_cluster": "00_core.ipynb",
         "Point": "00_core.ipynb",
         "pydot.Graph.__add__": "00_core.ipynb",
         "pydot.Node.__rshift__": "00_core.ipynb",
         "pydot.Graph.__rshift__": "00_core.ipynb",
         "is_listy": "00_core.ipynb"}

modules = ["core.py"]

doc_url = "https://fastai.github.io/fastdot/"

git_url = "https://github.com/fastai/fastdot/tree/master/"

def custom_doc_links(name): return None

fastdot/drawgraph.py
#AUTOGENERATED! DO NOT EDIT! File to edit: dev/00_drawgraph.ipynb (unless otherwise specified).

__all__ = ['pydot', 'Dot', 'quote', 'node_defaults', 'Node', 'obj2node_color', 'graph_colors1', 'graph_colors2',
           'Cluster', 'seq_cluster']

#Cell
from fastcore.all import *

import pydot
from matplotlib.colors import rgb2hex, hex2color

#Cell
def Dot(defaults=None, rankdir='LR', directed=True, compound=True, **kwargs):
    "Create a `pydot.Dot` graph with fastai/fastdot style defaults"
    return pydot.Dot(rankdir=rankdir, directed=directed, compound=compound, **kwargs)

#Cell
def _uniq_name(): return 'n'+(uuid4().hex)

def quote(x, q='"'):
    'Surround `x` with `"`'
    return f'"{x}"'

@patch
def _repr_svg_(self:pydot.Dot):
    return self.create_svg().decode('utf-8')

#Cell
def _pydot_create(f, label, **kwargs):
    kwargs = {k:v(label) if callable(v) else quote(v) for k,v in kwargs.items()}
    return f(_uniq_name(), label=quote(label), **kwargs)

#Cell
node_defaults = dict(shape='box', style='rounded, filled', fillcolor='white')

#Cell
def Node(label, tooltip=None, **kwargs):
    "Create a `pydot.Node` with a unique name"
    if not isinstance(label,str) and isinstance(label, Collection) and len(label)==2:
        label,tooltip = label
    kwargs = merge(node_defaults, kwargs)
    if tooltip is None: tooltip = label
    return _pydot_create(pydot.Node, label, tooltip=tooltip, **kwargs)

#Cell
def obj2node_color(cm, minalpha, rangealpha, o):
    "Create a consistent mapping from objects to colors, using colormap `cm`"
    h = hash(o)
    i = float(h % 256) / 256
    alpha = (h^hash('something')) % rangealpha + minalpha
    return rgb2hex(cm(i)) + f'{alpha:02X}'

#Cell
graph_colors1 = partial(obj2node_color, plt.get_cmap('rainbow'), 30, 160)
graph_colors2 = partial(obj2node_color, plt.get_cmap('tab20'), 30, 160)

#Cell
node_defaults['fillcolor'] = graph_colors1

#Cell
def Cluster(label='', tooltip=None, style='rounded, filled', fillcolor='#55555522', **kwargs):
    "Create a `pydot.Cluster` with a unique name"
    if tooltip is None: tooltip = label
    return _pydot_create(pydot.Cluster, label, tooltip=tooltip, style=style, fillcolor=fillcolor, **kwargs)

#Cell
@patch
def nodes(self:pydot.Graph):
    "`i`th node in `Graph`"
    return L(o for o in self.get_nodes() if o.get_label() is not None)

#Cell
@patch
def __getitem__(self:pydot.Graph, i):
    "`i`th node in `Graph`"
    return self.nodes()[i]

#Cell
@patch
def add_item(self:pydot.Graph, item, **kwargs):
    "Add a `Cluster`, `Node`, or `Edge` to the `Graph`"
    if not isinstance(item, (pydot.Edge,pydot.Node,pydot.Graph)): item = Node(item, **kwargs)
    f = self.add_node     if isinstance(item, pydot.Node ) else \
        self.add_subgraph if isinstance(item, pydot.Graph) else \
        self.add_edge     if isinstance(item, pydot.Edge ) else None
    f(item)
    return item

#Cell
@patch
def add_items(self:pydot.Graph, *items, **kwargs):
    "Add `items` the `Graph`"
    return L(self.add_item(it, **kwargs) for it in items)

#Cell
@patch
def first(self:pydot.Graph):
    "First node in `Graph`, searching subgraphs recursively as needed"
    nodes = self.nodes()
    if nodes: return nodes[0]
    for subg in self.get_subgraphs():
        res = subg.first()
        if res: return res

#Cell
@patch
def last(self:pydot.Graph):
    "Lastt node in `Graph`, searching subgraphs recursively as needed"
    nodes = self.nodes()
    if nodes: return nodes[-1]
    for subg in reversed(self.get_subgraphs()):
        res = subg.last()
        if res: return res

#Cell
@patch
def connect(self:(pydot.Node,pydot.Graph), item):
    "Connect two nodes or clusters"
    a,b,ltail,lhead = self,item,'',''
    if isinstance(self,pydot.Graph):
        a = self.last()
        ltail=self.get_name()
    if isinstance(item,pydot.Graph):
        b = item.first()
        lhead=item.get_name()
    return pydot.Edge(a, b, lhead=lhead, ltail=ltail)

#Cell
@patch
def seq_edges(self:pydot.Graph, items):
    "Add edges between each pair of nodes in `items`"
    return L(items[i].connect(items[i+1]) for i in range(len(items)-1))

#Cell
def seq_cluster(items, cluster_label='', **kwargs):
    sg = Cluster(cluster_label)
    its = sg.add_items(*items, **kwargs)
    sg.add_items(*sg.seq_edges(its))
    return sg

fastgraph/pagegraph.py
#AUTOGENERATED! DO NOT EDIT! File to edit: dev/00_pagegraph.ipynb (unless otherwise specified).

__all__ = ['split', 'make_dot_graph', 'to_dict', 'DotGraph', 'draw_graph', 'iter_nodes', 'map_', 'pfx',
           'external_inputs', 'bindings', 'walk', 'to_graph', 'to_graph', 'maybe_graph', 'explode', 'ColorMap',
           'Network', 'to_network']

#Cell
from functools import singledispatch
from torch import nn

#Cell
def split(path, sep = '/'):
    i = path.rfind(sep) + 1
    return path[:i].rstrip(sep), path[i:]

#Cell
def make_dot_graph(nodes, edges, direction='LR', **kwargs):
    from pydot import Dot, Cluster, Node, Edge
    class Subgraphs(dict):
        def __missing__(self, path):
            parent, label = split(path)
            subgraph = Cluster(path, label=label, style='rounded, filled', fillcolor='#77777744')
            self[parent].add_subgraph(subgraph)
            return subgraph
    g = Dot(rankdir=direction, directed=True, **kwargs)
    g.set_node_defaults(
        shape='box', style='rounded, filled', fillcolor='#ffffff')
    subgraphs = Subgraphs({'': g})
    for path, attr in nodes:
        parent, label = split(path)
        subgraphs[parent].add_node(
            Node(name=path, label=label, **attr))
    for src, dst, attr in edges:
        g.add_edge(Edge(src, dst, **attr))
    return g

#Cell
def to_dict(inputs):
    return dict(enumerate(inputs)) if isinstance(inputs, list) else inputs

#Cell
class DotGraph():
    def __init__(self, graph, size=None, direction='LR'):
        self.nodes = [(k, v) for k, (v,_) in graph.items()]
        self.edges = [(src, dst, {'tooltip': name}) for dst, (_, inputs) in graph.items() for name, src in to_dict(inputs).items()]
        self.size, self.direction = size or 8+len(graph)/3, direction

    def dot_graph(self, **kwargs):
        return make_dot_graph(self.nodes, self.edges, size=self.size, direction=self.direction,  **kwargs)

    def svg(self, **kwargs):
        return self.dot_graph(**kwargs).create(format='svg').decode('utf-8')
    try:
        import pydot
        _repr_svg_ = svg
    except ImportError:
        def __repr__(self): return 'pydot is needed for network visualisation'

#Cell
def draw_graph(graph):
    return DotGraph({name: ({'tooltip': str(value)}, inputs) for name, (value,inputs) in graph.items()}, size=10)

#Cell
def iter_nodes(graph):
    # graph = {name: node for (name, node) in graph.items() if node is not None}
    keys = list(graph.keys())
    if not all(isinstance(k, str) for k in keys):
        raise Exception('Node names must be strings.')
    return ((name, (node if isinstance(node, tuple) else (node, [0 if j is 0 else keys[j-1]])))
            for (j, (name, node)) in enumerate(graph.items()))

#Cell
map_ = lambda func, vals: [func(x) for x in vals] if isinstance(vals, list) else {k: func(v) for k,v in vals.items()}
pfx = lambda prefix, name: f'{prefix}/{name}'
external_inputs = lambda graph: set(i for name, (value, inputs) in iter_nodes(graph) for i in inputs if i not in graph)

#Cell
def bindings(graph, inputs):
    if isinstance(inputs, list): inputs = dict(enumerate(inputs))
    required_inputs = external_inputs(graph)
    missing = [k for k in required_inputs if k not in inputs]
    if len(missing):
        raise Exception(f'Required inputs {missing} are missing from inputs {inputs} for graph {graph}')
    return inputs

#Cell
walk = lambda dct, key: walk(dct, dct[key]) if key in dct else key

#Cell
@singledispatch
def to_graph(value):
    raise NotImplementedError(f'type = {type(value)}')

#Cell
@to_graph.register(dict)
def _(x): return x

#Cell
def maybe_graph(x):
    try: return to_graph(x)
    except NotImplementedError: return x

#Cell
def explode(graph, max_levels=-1, convert=maybe_graph):
    graph = convert(graph)
    if max_levels==0 or not isinstance(graph, dict): return graph
    redirects = {}
    def iter_(graph):
        for name, (value, inputs) in iter_nodes(graph):
            value = explode(value, max_levels-1, convert=convert)
            if isinstance(value, dict):
                #special case empty dict
                if not len(value):
                    if len(inputs) != 1: raise Exception('Empty graphs (pass-thrus) should have exactly one input')
                    redirects[name] = inputs[0] #redirect to input
                else:
                    bindings_dict = bindings(value, inputs)
                    for n, (val, ins) in iter_nodes(value):
                        yield (pfx(name, n), (val, map_((lambda i: bindings_dict.get(i, pfx(name, i))), ins)))
                    redirects[name] = pfx(name, n) #redirect to previous node
            else:
                yield (name, (value, inputs))
    return {name: (value, map_((lambda i: walk(redirects, i)), inputs)) for name, (value, inputs) in iter_(graph)}

#Cell
class ColorMap(dict):
    palette = ['#'+x for x in (
        'bebada,ffffb3,fb8072,8dd3c7,80b1d3,fdb462,b3de69,fccde5,bc80bd,ccebc5,ffed6f,1f78b4,33a02c,e31a1c,ff7f00,'
        '4dddf8,e66493,b07b87,4e90e3,dea05e,d0c281,f0e189,e9e8b1,e0eb71,bbd2a4,6ed641,57eb9c,3ca4d4,92d5e7,b15928'
    ).split(',')]

    def __missing__(self, key):
        self[key] = self.palette[len(self) % len(self.palette)]
        return self[key]

    def _repr_html_(self):
        css = (
        '.pill {'
            'margin:2px; border-width:1px; border-radius:9px; border-style:solid;'
            'display:inline-block; width:100px; height:15px; line-height:15px;'
        '}'
        '.pill_text {'
            'width:90%; margin:auto; font-size:9px; text-align:center; overflow:hidden;'
        '}'
        )
        s = '<div class=pill style="background-color:{}"><div class=pill_text>{}</div></div>'
        return '<style>'+css+'</style>'+''.join((s.format(color, text) for text, color in self.items()))

#Cell
class Network(nn.Module):
    colors = ColorMap()
    def __init__(self, graph, cache_activations=False):
        self._graph = to_graph(graph)
        super().__init__()
        self.cache_activations = cache_activations
        for path, (val, _) in iter_nodes(self._graph):  setattr(self, path.replace('/', '__'), val)

    def __setattr__(self, name, value):
        super().__setattr__(name, value)
        path = name.replace('__', '/')
        if path in self._graph:
            old_val = self._graph[path]
            self._graph[path] = (value, old_val[1]) if isinstance(old_val, tuple) else value

    def forward(self, *args):
        prev = args[0]
        outputs = self.cache = dict(enumerate(args))
        for k, (node, inputs) in iter_nodes(self._graph):
            if k not in outputs:prev = outputs[k] = node(*[outputs[x] for x in inputs])
        if not self.cache_activations: self.cache = None
        return prev

    def draw(self, **kwargs):
        return DotGraph({p: ({'fillcolor': self.colors[type(v).__name__], 'tooltip': str(v)}, inputs)
                         for p, (v, inputs) in iter_nodes(to_graph(self))}, **kwargs)

    def explode(self, max_levels=-1):
        convert = lambda x: to_graph(x) if isinstance(x, Network) else x
        return Network(explode(self, max_levels, convert=convert))

#Cell
def to_network(module, max_levels=-1):
    net = Network(module)
    if max_levels == 0: return net
    for k, mod in net.named_children():
        try: setattr(net, k, to_network(mod, max_levels-1))
        except NotImplementedError: pass
    return net

#Cell
@to_graph.register(Network)
def _(x): return x._graph

fastdot/core.py
# AUTOGENERATED! DO NOT EDIT! File to edit: 00_core.ipynb (unless otherwise specified).

__all__ = ['pydot', 'Dot', 'uniq_name', 'quote', 'graph_objects', 'object_names', 'add_mapping', 'node_defaults',
           'Node', 'object2graph', 'obj2node_color', 'graph_colors1', 'graph_colors2', 'cluster_defaults', 'Cluster',
           'graph_items', 'object_connections', 'graph_edges_seq', 'seq_cluster', 'Point', 'is_listy']

# Cell
from fastcore.all import *
import pydot, matplotlib.pyplot as plt
from matplotlib.colors import rgb2hex, hex2color
from uuid import uuid4
from typing import Collection

# Cell
#nbdev_comment _all_ = ['pydot']

# Cell
def Dot(defaults=None, rankdir='LR', directed=True, compound=True, **kwargs):
    "Create a `pydot.Dot` graph with fastai/fastdot style defaults"
    return pydot.Dot(rankdir=rankdir, directed=directed, compound=compound, **kwargs)

# Cell
def uniq_name(o): return 'n'+(uuid4().hex)

def quote(x, q='"'):
    'Surround `x` with `"`'
    return f'"{x}"'

@patch
def _repr_svg_(self:pydot.Dot):
    return self.create_svg().decode('utf-8')

# Cell
graph_objects = {}
object_names = {}

# Cell
def add_mapping(graph_item, obj):
    graph_objects[graph_item.get_name()] = graph_item
    object_names[id(obj)] = graph_item.get_name()
    return graph_item

# Cell
def _pydot_create(f, obj, **kwargs):
    for k,v in kwargs.items():
        if callable(v): v = kwargs[k] = v(obj)
        if k not in ('name','graph_name'): kwargs[k] = quote(v)
    return add_mapping(f(**kwargs), obj)

# Cell
node_defaults = dict(label=str, tooltip=str, name=uniq_name, shape='box', style='rounded, filled', fillcolor='white')

# Cell
def Node(obj, **kwargs):
    "Create a `pydot.Node` with a unique name"
    if not isinstance(obj,str) and isinstance(obj, Collection) and len(obj)==2:
        obj,kwargs['tooltip'] = obj
    kwargs = merge(node_defaults, kwargs)
    return _pydot_create(pydot.Node, obj, **kwargs)

# Cell
def object2graph(o):
    "Get graph item representing `o`"
    return graph_objects[object_names[id(o)]]

# Cell
def obj2node_color(cm, minalpha, rangealpha, o):
    "Create a consistent mapping from objects to colors, using colormap `cm`"
    h = hash(o)
    i = float(h % 256) / 256
    alpha = (h^hash('something')) % rangealpha + minalpha
    return rgb2hex(cm(i)) + f'{alpha:02X}'

# Cell
graph_colors1 = partial(obj2node_color, plt.get_cmap('rainbow'), 30, 160)
graph_colors2 = partial(obj2node_color, plt.get_cmap('tab20'), 30, 160)

# Cell
node_defaults['fillcolor'] = graph_colors1

# Cell
cluster_defaults = dict(label=str, tooltip=str, graph_name=uniq_name, style='rounded, filled', fillcolor='#55555522')

# Cell
def Cluster(obj='', **kwargs):
    "Create a `pydot.Cluster` with a unique name"
    kwargs = merge(cluster_defaults, kwargs)
    return _pydot_create(pydot.Cluster, obj, **kwargs)

# Cell
@patch
def nodes(self:pydot.Graph):
    "`i`th node in `Graph`"
    return L(o for o in self.get_nodes() if o.get_label() is not None)

# Cell
@patch
def __getitem__(self:pydot.Graph, i):
    "`i`th node in `Graph`"
    return self.nodes()[i]

# Cell
@patch
def add_item(self:pydot.Graph, item, **kwargs):
    "Add a `Cluster`, `Node`, or `Edge` to the `Graph`"
    if not isinstance(item, (pydot.Edge,pydot.Node,pydot.Graph)): item = Node(item, **kwargs)
    f = self.add_node     if isinstance(item, pydot.Node ) else \
        self.add_subgraph if isinstance(item, pydot.Graph) else \
        self.add_edge     if isinstance(item, pydot.Edge ) else None
    f(item)
    return item

# Cell
@patch
def add_items(self:pydot.Graph, *items, **kwargs):
    "Add `items` the `Graph`"
    return L(self.add_item(it, **kwargs) for it in items)

# Cell
def graph_items(*items, **kwargs):
    "Add `items` to a new `pydot.Dot`"
    g = Dot()
    g.add_items(*items, **kwargs)
    return g

# Cell
@patch
def first(self:pydot.Graph):
    "First node in `Graph`, searching subgraphs recursively as needed"
    nodes = self.nodes()
    if nodes: return nodes[0]
    for subg in self.get_subgraphs():
        res = subg.first()
        if res: return res

# Cell
@patch
def last(self:pydot.Graph):
    "Lastt node in `Graph`, searching subgraphs recursively as needed"
    nodes = self.nodes()
    if nodes: return nodes[-1]
    for subg in reversed(self.get_subgraphs()):
        res = subg.last()
        if res: return res

# Cell
@patch
def with_compass(self:(pydot.Node,pydot.Graph), compass=None):
    r = self.get_name()
    return f'{r}:{compass}' if compass else r

# Cell
@patch
def connect(self:(pydot.Node,pydot.Graph), item, compass1=None, compass2=None, **kwargs):
    "Connect two nodes or clusters"
    a,b,ltail,lhead = self,item,'',''
    if isinstance(self,pydot.Graph):
        a = self.last()
        ltail=self.get_name()
    if isinstance(item,pydot.Graph):
        b = item.first()
        lhead=item.get_name()
    a,b = a.with_compass(compass1),b.with_compass(compass2)
    return pydot.Edge(a, b, lhead=lhead, ltail=ltail, **kwargs)

# Cell
def object_connections(conns):
    "Create connections between all pairs in `conns`"
    return [object2graph(a).connect(object2graph(b)) for a,b in conns]

# Cell
def graph_edges_seq(items):
    "Add edges between each pair of nodes in `items`"
    return L(items[i].connect(items[i+1]) for i in range(len(items)-1))

# Cell
@patch
def add_edges_seq(self:pydot.Graph, items):
    "Add edges between each pair of nodes in `items`"
    return self.add_items(*graph_edges_seq(items))

# Cell
def seq_cluster(items, cluster_label='', **kwargs):
    sg = Cluster(cluster_label)
    its = sg.add_items(*items, **kwargs)
    sg.add_edges_seq(its)
    return sg

# Cell
def Point(label='pnt', **kwargs):
    "Create a `Node` with a 'point' shape"
    return (Node('pnt', shape='point'))

# Cell
@patch
def __add__(self:pydot.Graph, item, **kwargs):
    "Add a `Cluster`, `Node`, or `Edge` to the `Graph` with operator"
    # if not is_listy(item):
    #     item = L(item)
    if not is_listy(item):
        return self.add_item(item, **kwargs)
    else:
        return self.add_items( *item, **kwargs)

@patch
def __rshift__(self:(pydot.Node,pydot.Graph), item, compass1=None, compass2=None, **kwargs):
    return self.connect(item, compass1, compass2, **kwargs)


def is_listy(x)->bool:
    return isinstance(x, (tuple,list))